compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -1076,6 +1076,21 @@ pub struct VarDebugInfo<'tcx> {
 // BasicBlock
 
 rustc_index::newtype_index! {
+    /// The unit of the MIR [control-flow graph][CFG].
+    ///
+    /// There is no branching (e.g., `if`s, function calls, etc.) within a basic block, which makes
+    /// it easier to do [data-flow analyses] and optimizations. Instead, branches are represented
```

Comment:
I use the form with the hyphen here since that seems to be the canonical way: https://en.wikipedia.org/wiki/Data-flow_analysis. Perhaps the rustc-dev-guide should be changed?

Reply:
If the rustc-dev-guide uses a different hyphenization, then yeah it probably should be changed.

Reply:
Actually it seems like it's referred to as "dataflow analysis" (without the hyphen) in a bunch of places in the compiler:

<details>
<summary>grep output</summary>
<pre>
compiler/rustc_data_structures/src/graph/implementation/mod.rs://! A graph module for use in dataflow, region resolution, and elsewhere.
compiler/rustc_passes/src/liveness.rs://! A classic liveness analysis based on dataflow over the AST. Computes,
compiler/rustc_middle/src/ty/adjustment.rs:/// is implemented. Right now, dataflow can't handle the general case where there
compiler/rustc_middle/src/mir/mod.rs:    ///     https://rustc-dev-guide.rust-lang.org/appendix/background.html#what-is-a-dataflow-analysis
compiler/rustc_mir/src/dataflow/framework/direction.rs:                        // Propagate dataflow state along the "otherwise" edge.
compiler/rustc_mir/src/dataflow/framework/graphviz.rs://! A helpful diagram for debugging dataflow problems.
compiler/rustc_mir/src/dataflow/framework/graphviz.rs:        // Write the full dataflow state immediately after the terminator if it differs from the
compiler/rustc_mir/src/dataflow/framework/graphviz.rs:        // backward dataflow analyses as well as effects for `SwitchInt` terminators.
compiler/rustc_mir/src/dataflow/framework/fmt.rs://! Custom formatting traits used when outputting Graphviz diagrams with the results of a dataflow
compiler/rustc_mir/src/dataflow/framework/mod.rs://! A framework that can express both [gen-kill] and generic dataflow problems.
compiler/rustc_mir/src/dataflow/framework/mod.rs://! `impls` module contains several examples of gen/kill dataflow analyses.
compiler/rustc_mir/src/dataflow/framework/mod.rs://! fixpoint solution to your dataflow problem, or implement the `ResultsVisitor` interface and use
compiler/rustc_mir/src/dataflow/framework/mod.rs://! use rustc_mir::dataflow::Analysis; // Makes `into_engine` available.
compiler/rustc_mir/src/dataflow/framework/mod.rs://!     // Print the dataflow state *after* each statement in the start block.
compiler/rustc_mir/src/dataflow/framework/mod.rs:/// Define the domain of a dataflow problem.
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// The type that holds the dataflow state at any given point in the program.
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// The initial value of the dataflow state upon entry to each basic block.
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Mutates the initial value of the dataflow state upon entry to the `START_BLOCK`.
compiler/rustc_mir/src/dataflow/framework/mod.rs:    // FIXME: For backward dataflow analyses, the initial state should be applied to every basic
compiler/rustc_mir/src/dataflow/framework/mod.rs:/// A dataflow problem with an arbitrarily complex transfer function.
compiler/rustc_mir/src/dataflow/framework/mod.rs:/// > If the dataflow state **before** some point in the program changes to be greater
compiler/rustc_mir/src/dataflow/framework/mod.rs:/// than the prior state **before** that point, the dataflow state **after** that point must
compiler/rustc_mir/src/dataflow/framework/mod.rs:/// This invariant guarantees that the dataflow state at a given point in the program increases
compiler/rustc_mir/src/dataflow/framework/mod.rs:/// to the same point in the program at different points in time. The dataflow state at a given
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Updates the current dataflow state with the effect of evaluating a statement.
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Updates the current dataflow state with an effect that occurs immediately *before* the
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Updates the current dataflow state with the effect of evaluating a terminator.
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Updates the current dataflow state with an effect that occurs immediately *before* the
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Updates the current dataflow state with the effect of a successful return from a `Call`
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Updates the current dataflow state with the effect of resuming from a `Yield` terminator.
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Updates the current dataflow state with the effect of taking a particular branch in a
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// FIXME: This class of effects is not supported for backward dataflow analyses.
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Creates an `Engine` to find the fixpoint for this dataflow problem.
compiler/rustc_mir/src/dataflow/framework/mod.rs:/// A gen/kill dataflow problem.
compiler/rustc_mir/src/dataflow/framework/mod.rs:/// allow modification of the dataflow state via "gen" and "kill" operations. By defining transfer
compiler/rustc_mir/src/dataflow/framework/mod.rs:    /// Creates a new transfer function that will leave the dataflow state unchanged.
compiler/rustc_mir/src/dataflow/framework/cursor.rs://! Random access inspection of the results of a dataflow analysis.
compiler/rustc_mir/src/dataflow/framework/cursor.rs:/// Allows random access inspection of the results of a dataflow analysis.
compiler/rustc_mir/src/dataflow/framework/cursor.rs:/// A `ResultsCursor` can either own (the default) or borrow the dataflow results it inspects. The
compiler/rustc_mir/src/dataflow/framework/cursor.rs:    /// Returns the dataflow state at the current location.
compiler/rustc_mir/src/dataflow/framework/cursor.rs:    /// For forward dataflow analyses, this is the dataflow state prior to the first statement.
compiler/rustc_mir/src/dataflow/framework/cursor.rs:    /// For backward dataflow analyses, this is the dataflow state after the terminator.
compiler/rustc_mir/src/dataflow/framework/cursor.rs:    /// Advances the cursor to hold the dataflow state at `target` before its "primary" effect is
compiler/rustc_mir/src/dataflow/framework/cursor.rs:    /// Advances the cursor to hold the dataflow state at `target` after its "primary" effect is
compiler/rustc_mir/src/dataflow/framework/cursor.rs:    /// creating an extra copy of the dataflow state.
compiler/rustc_mir/src/dataflow/framework/tests.rs:/// A dataflow analysis whose state is unique at every possible `SeekTarget`.
compiler/rustc_mir/src/dataflow/framework/tests.rs:    /// Returns the index that should be added to the dataflow state at the given target.
compiler/rustc_mir/src/dataflow/framework/lattice.rs://! Traits used to represent [lattices] for use as the domain of a dataflow analysis.
compiler/rustc_mir/src/dataflow/framework/lattice.rs://! dataflow analysis, there's no need for a newtype wrapper with a custom `PartialOrd` impl. The
compiler/rustc_mir/src/dataflow/framework/lattice.rs:/// `MeetSemiLattice`. However, types that will be used as dataflow domains should implement both
compiler/rustc_mir/src/dataflow/framework/visitor.rs:/// dataflow state at that location.
compiler/rustc_mir/src/dataflow/framework/visitor.rs:/// This trait exists so that we can visit the results of multiple dataflow analyses simultaneously.
compiler/rustc_mir/src/dataflow/framework/visitor.rs:    /// Creates an empty `FlowState` to hold the transient state for these dataflow results.
compiler/rustc_mir/src/dataflow/framework/visitor.rs:/// dataflow analyses used by the borrow checker.
compiler/rustc_mir/src/dataflow/framework/visitor.rs:/// The results of the dataflow analyses used by the borrow checker.
compiler/rustc_mir/src/dataflow/framework/visitor.rs:/// The transient state of the dataflow analyses used by the borrow checker.
compiler/rustc_mir/src/dataflow/framework/engine.rs://! A solver for dataflow problems.
compiler/rustc_mir/src/dataflow/framework/engine.rs:/// A dataflow analysis that has converged to fixpoint.
compiler/rustc_mir/src/dataflow/framework/engine.rs:    /// Gets the dataflow state for the given block.
compiler/rustc_mir/src/dataflow/framework/engine.rs:/// A solver for dataflow problems.
compiler/rustc_mir/src/dataflow/framework/engine.rs:    /// Creates a new `Engine` to solve a gen-kill dataflow problem.
compiler/rustc_mir/src/dataflow/framework/engine.rs:    /// Creates a new `Engine` to solve a dataflow problem with an arbitrary transfer
compiler/rustc_mir/src/dataflow/framework/engine.rs:    /// Signals that we do not want dataflow state to propagate across unwind edges for these
compiler/rustc_mir/src/dataflow/framework/engine.rs:    /// unwind during execution. Otherwise, your dataflow results will not be correct.
compiler/rustc_mir/src/dataflow/framework/engine.rs:    /// Computes the fixpoint for this dataflow problem and returns it.
compiler/rustc_mir/src/dataflow/framework/engine.rs:            // backward dataflow analyses, but probably not enough to matter.
compiler/rustc_mir/src/dataflow/framework/engine.rs:/// Writes a DOT file containing the results of a dataflow analysis if the user requested it via
compiler/rustc_mir/src/dataflow/framework/engine.rs:    /// Returns the path where dataflow results should be written, or `None`
compiler/rustc_mir/src/dataflow/impls/liveness.rs:/// A [live-variable dataflow analysis][liveness].
compiler/rustc_mir/src/dataflow/impls/liveness.rs:/// [flow-test]: https://github.com/rust-lang/rust/blob/a08c47310c7d49cbdc5d7afb38408ba519967ecd/src/test/ui/mir-dataflow/liveness-ptr.rs
compiler/rustc_mir/src/dataflow/impls/mod.rs:/// dataflow information shown in the right-hand comments.
compiler/rustc_mir/src/dataflow/impls/mod.rs:/// dataflow information shown in the right-hand comments.
compiler/rustc_mir/src/dataflow/impls/mod.rs:/// dataflow information shown in the right-hand comments.
compiler/rustc_mir/src/dataflow/impls/mod.rs:/// This dataflow is used to determine if an immutable local variable may
compiler/rustc_mir/src/dataflow/impls/mod.rs:/// dataflow information shown in the right-hand comments.
compiler/rustc_mir/src/dataflow/impls/borrowed_locals.rs:/// A dataflow analysis that tracks whether a pointer or reference could possibly exist that points
compiler/rustc_mir/src/dataflow/impls/borrowed_locals.rs:    /// A dataflow analysis that records whether a pointer or reference exists that may alias the
compiler/rustc_mir/src/dataflow/impls/borrowed_locals.rs:    /// A dataflow analysis that records whether a pointer or reference exists that may *mutably*
compiler/rustc_mir/src/dataflow/impls/borrowed_locals.rs:    /// During dataflow analysis, ignore the borrow that may occur when a place is dropped.
compiler/rustc_mir/src/dataflow/impls/borrows.rs:        // reflects the dataflow on entry to the statement.
compiler/rustc_mir/src/dataflow/move_paths/mod.rs:/// With respect to dataflow analysis:
compiler/rustc_mir/src/borrow_check/places_conflict.rs:/// dataflow).
compiler/rustc_mir/src/borrow_check/mod.rs:        // The dataflow tracks shallow prefixes distinctly (that is,
compiler/rustc_mir/src/borrow_check/mod.rs:        // ancestors; dataflow recurs on children when parents
compiler/rustc_mir/src/borrow_check/type_check/liveness/trace.rs:    /// Results of dataflow tracking which variables (and paths) have been
compiler/rustc_mir/src/borrow_check/type_check/mod.rs:/// - `flow_inits` -- results of a maybe-init dataflow analysis
compiler/rustc_mir/src/borrow_check/type_check/mod.rs:/// - `move_data` -- move-data constructed when performing the maybe-init dataflow analysis
compiler/rustc_mir/src/transform/check_consts/validation.rs:            // need to run dataflow.
compiler/rustc_mir/src/transform/check_consts/resolver.rs://! This contains the dataflow analysis used to track `Qualif`s on complex control-flow graphs.
compiler/rustc_mir/src/transform/check_consts/resolver.rs:/// the `MaybeMutBorrowedLocals` dataflow pass to see if a `Local` may have become qualified via
compiler/rustc_mir/src/transform/check_consts/resolver.rs:/// The dataflow analysis used to propagate qualifs on arbitrary CFGs.
compiler/rustc_mir/src/transform/check_consts/qualifs.rs:    /// The name of the file used to debug the dataflow analysis that computes this qualif.
compiler/rustc_mir/src/transform/rustc_peek.rs:/// For each such call, determines what the dataflow bit-state is for
compiler/rustc_mir/src/transform/rustc_peek.rs:/// The intention is that one can write unit tests for dataflow by
compiler/rustc_mir/src/transform/rustc_peek.rs:/// make observations about the results of dataflow static analyses.
</pre>
</details>

So maybe it would be better to leave the rustc-dev-guide and the other references as they are from a pragmatic sense.

Reply:
No strong opinion from my side.